pkg/controller/utils/utils.go

```diff
@@ -76,3 +76,25 @@ func GetCredentialsRequestCloudType(providerSpec *runtime.RawExtension) (string,
 	}
 	return unknown.Kind, nil
 }
+
+// GenerateNameWithFieldLimits lets you pass in two strings which will be clipped to their respective
+// maximum lengths.
+// Example: passing "thisIsInfraName", 8, "thisIsCrName", 8 will return:
+//      'thisIsIn-thisIsCr'
+func GenerateNameWithFieldLimits(infraName string, infraNameMaxLen int, crName string, crNameLen int) (string, error) {
+	if crName == "" {
+		return "", fmt.Errorf("empty credential request name")
+	}
+
+	infraPrefix := ""
+	if infraName != "" {
+		if len(infraName) > infraNameMaxLen {
+			infraName = infraName[0:infraNameMaxLen]
+		}
+		infraPrefix = infraName + "-"
+	}
+	if len(crName) > crNameLen {
+		crName = crName[0:crNameLen]
```

Comment:
I would still error when the entire generated name is longer than 93 instead of truncating the string. One might create two requests where both request name has first 60 characters in common. In which case both requests will get the same service principal. And if both requests have different list of roles, one request will override roles of another request.

Reply:
In that case we will first need to update CCR status [1] and set the generated service principal name there [2] before the AAD application is created.

[1] https://github.com/openshift/cloud-credential-operator/blob/master/pkg/azure/actuator.go#L426-L429
[2] https://github.com/openshift/cloud-credential-operator/blob/master/pkg/azure/actuator.go#L372

Reply:
> rather than save the generated names of things, couldn't we just find the App Registration and the Service Principal by their appID fields? we already save the appID in the CCR status.

There is a risk you will be able to create AAD application but fail to update CCR status and set the appID there. In which case you will generate different app name and will have orphaned app application resource. The resource will have no roles assigned, no secret generated so it will cause no harm. As long as that's fine, even this makes sense.